modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+            AionBlock first = blockRange.remove(0);
+
+            int stored = addBlockRange(first, blockRange);
+
+            // save data to disk
+            indexSource.commitBatch();
+            levelSource.flushBatch();
+            queueSource.flushBatch();
+
+            // the number of blocks added
+            return stored;
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /** Stores a block ranges with the first element determining the queue placement. */
+    private int addBlockRange(AionBlock first, List<AionBlock> blockRange) {
+
+        // skip if already stored
+        while (indexSource.get(first.getHash()).isPresent()) {
+            if (blockRange.size() == 0) {
+                return 0;
+            } else {
+                first = blockRange.remove(0);
+            }
+        }
+
+        // the first block is not stored
+        // start new queue with hash = first node hash
+        byte[] currentQueueHash = first.getHash();
+        List<AionBlock> currentQueue = new ArrayList<>();
+
+        // add (to) level
+        byte[] levelKey = ByteUtil.longToBytes(first.getNumber());
+        List<byte[]> levelData = levelSource.get(levelKey);
+
+        if (levelData == null) {
+            levelData = new ArrayList<>();
+        }
+
+        levelData.add(currentQueueHash);
+        levelSource.putToBatch(levelKey, levelData);
+
+        // index block with queue hash
+        indexSource.putToBatch(first.getHash(), currentQueueHash);
+        int stored = 1;
+
+        // add element to queue
+        currentQueue.add(first);
+
+        // keep track of parent to ensure correct range
+        AionBlock parent = first;
+
+        AionBlock current;
+        // process rest of block range
+        while (blockRange.size() > 0) {
+            current = blockRange.remove(0);
+
+            // check for issues with batch continuity and storage
+            if (!Arrays.equals(current.getParentHash(), parent.getHash()) // continuity issue
+                    || indexSource.get(current.getHash()).isPresent()) { // already stored
+
+                // store separately
+                stored += addBlockRange(current, blockRange);
+
+                // done with loop
+                break;
+            }
+
+            // index block to current queue
+            indexSource.putToBatch(current.getHash(), currentQueueHash);
+
+            // append block to queue
+            currentQueue.add(current);
+            stored++;
+
+            // update parent
+            parent = current;
+        }
+
+        // done with queue
+        queueSource.putToBatch(currentQueueHash, currentQueue);
+
+        // the number of blocks added
+        return stored;
+    }
+
+    /**
+     * @return the number of elements stored in the index database.
+     * @implNote This method is package private because it is meant to be used for testing.
```

Comment:
how about put @VisibleForTesting in front of this method?

Reply:
updated in 2bc2919